### PR TITLE
Handle resolve request for emitter/flusher.

### PR DESCRIPTION
**Pull request by Benzbeeb**

### CODE REVIEW

File: chain/emitter/handler.go [hunk 1]
```diff
@@ -23,3 +27,49 @@ func (app *App) handleMsg(txHash []byte, msg sdk.Msg, log sdk.ABCIMessageLog, ex
 		app.handleMsgReportData(txHash, msg, evMap, extra)
 	}
 }
+
+func (app *App) handleEndBlock(event abci.Event) {
+	kvMap := make(map[string]string)
```
Comment (taobun) [inline chain/emitter/handler.go]: Separate this function to `transformEvent`

File: chain/emitter/handler.go [hunk 2]
```diff
@@ -23,3 +27,49 @@ func (app *App) handleMsg(txHash []byte, msg sdk.Msg, log sdk.ABCIMessageLog, ex
...
+	for _, kv := range event.Attributes {
+		kvMap[string(kv.Key)] = string(kv.Value)
+	}
+	switch event.Type {
+	case types.EventTypeRequestExecute:
+		id, err := strconv.ParseInt(kvMap[types.AttributeKeyRequestID], 10, 64)
```
Comment (taobun) [inline chain/emitter/handler.go]: use atoi

File: flusher/flusher/db.py [hunk 3]
```diff
@@ -1,6 +1,23 @@
...
 import sqlalchemy as sa
+import enum
+
+
+class ResolveStatus(enum.Enum):
+    Open = 0
+    Success = 1
+    Failure = 2
+    Expired = 3
+    Unknown = 4
```
Comment (taobun) [inline flusher/flusher/db.py]: Remove it

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/emitter/handler.go [hunk 4]
```diff
@@ -2,24 +2,53 @@ package emitter
...
+	switch event.Type {
+	case types.EventTypeRequestExecute:
+		resolveStatus := types.ResolveStatus(atoi(evMap[types.EventTypeRequestExecute+"."+types.AttributeKeyResolveStatus][0]))
+		dict := JsDict{
+			"id":             atoi(evMap[types.EventTypeRequestExecute+"."+types.AttributeKeyRequestID][0]),
+			"request_time":   atoi(evMap[types.EventTypeRequestExecute+"."+types.AttributeKeyRequestTime][0]),
+			"resolve_time":   atoi(evMap[types.EventTypeRequestExecute+"."+types.AttributeKeyResolveTime][0]),
+			"resolve_status": resolveStatus,
+		}
+		app.Write("UPDATE_REQUEST", dict)
```
Comment (taobun) [inline chain/emitter/handler.go]: Why we cannot write this message after add result?

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/emitter/handler.go [hunk 5]
```diff
@@ -2,24 +2,53 @@ package emitter
...
 	case oracle.MsgReportData:
 		app.handleMsgReportData(txHash, msg, evMap, extra)
 	}
 }
+
+func (app *App) handleEndBlock(event abci.Event) {
+	events := sdk.StringifyEvents([]abci.Event{event})
+	evMap := parseEvents(events)
+	switch event.Type {
+	case types.EventTypeRequestExecute:
```
Comment (sorawit) [inline chain/emitter/handler.go]: Let's expose this in alias.go, then use it as `oracle.EventTypeRequestExecute`

File: chain/emitter/handler.go [hunk 6]
```diff
@@ -2,24 +2,53 @@ package emitter
...
+		resolveStatus := types.ResolveStatus(atoi(evMap[types.EventTypeRequestExecute+"."+types.AttributeKeyResolveStatus][0]))
```
Comment (sorawit) [inline chain/emitter/handler.go]: I prefer we bring this logic to `oracle.go`

File: chain/emitter/handler.go [hunk 7]
```diff
@@ -2,24 +2,53 @@ package emitter
...
+		dict := JsDict{
+			"id":             atoi(evMap[types.EventTypeRequestExecute+"."+types.AttributeKeyRequestID][0]),
+			"request_time":   atoi(evMap[types.EventTypeRequestExecute+"."+types.AttributeKeyRequestTime][0]),
```
Comment (sorawit) [inline chain/emitter/handler.go]: Let's get all information apart from `id` from calling into the keeper's `GetResult` function. I think it's safer.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/oracle/alias.go [hunk 8]
```diff
@@ -14,10 +14,11 @@ const (
...
 var (
-	NewKeeper     = keeper.NewKeeper
-	NewQuerier    = keeper.NewQuerier
-	ModuleCdc     = types.ModuleCdc
-	RegisterCodec = types.RegisterCodec
+	NewKeeper               = keeper.NewKeeper
+	NewQuerier              = keeper.NewQuerier
+	ModuleCdc               = types.ModuleCdc
+	RegisterCodec           = types.RegisterCodec
+	EventTypeRequestExecute = types.EventTypeRequestExecute
```
Comment (taobun) [inline chain/x/oracle/alias.go:22]: Add new line between RegisterCodec and Event.

File: chain/emitter/oracle.go [hunk 9]
```diff
@@ -64,3 +65,21 @@ func (app *App) handleMsgReportData(
...
+	dict := JsDict{
+		"id":             id,
+		"request_time":   atoi(evMap[types.EventTypeRequestExecute+"."+types.AttributeKeyRequestTime][0]),
+		"resolve_time":   atoi(evMap[types.EventTypeRequestExecute+"."+types.AttributeKeyResolveTime][0]),
+		"resolve_status": resolveStatus,
+	}
+	if resolveStatus == types.ResolveStatus_Success {
+		result := []byte(evMap[types.EventTypeRequestExecute+"."+types.AttributeKeyResult][0])
+		dict["result"] = result
+		app.OracleKeeper.GetResult(app.DeliverContext, id)
```
Comment (taobun) [inline chain/emitter/oracle.go]: GetResult?

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/emitter/oracle.go [hunk 10]
```diff
@@ -64,3 +65,20 @@ func (app *App) handleMsgReportData(
 		})
 	}
 }
+
+// handleRequestExecuteimplements emitter handler for EventRequestExecute.
+func (app *App) handleRequestExecute(evMap EvMap) {
```
Comment (sorawit) [inline chain/emitter/oracle.go]: handleEventRequestExecute

File: chain/emitter/oracle.go [hunk 11]
```diff
@@ -64,3 +65,20 @@ func (app *App) handleMsgReportData(
 		})
 	}
 }
+
+// handleRequestExecuteimplements emitter handler for EventRequestExecute.
```
Comment (sorawit) [inline chain/emitter/oracle.go]: Add space

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/emitter/handler.go (hunk 7 above)
Comment (sorawit) [reply chain/emitter/handler.go]: This is not resolved

File: chain/emitter/oracle.go [hunk 12]
```diff
@@ -64,3 +65,20 @@ func (app *App) handleMsgReportData(
...
+// handleEventRequestExecute implements emitter handler for EventRequestExecute.
+func (app *App) handleEventRequestExecute(evMap EvMap) {
+	id := types.RequestID(atoi(evMap[types.EventTypeRequestExecute+"."+types.AttributeKeyRequestID][0]))
+	result := app.OracleKeeper.MustGetResult(app.DeliverContext, id)
+	resolveStatus := result.ResponsePacketData.ResolveStatus
+	dict := JsDict{
+		"id":             id,
+		"request_time":   result.ResponsePacketData.RequestTime,
+		"resolve_time":   result.ResponsePacketData.ResolveTime,
+		"resolve_status": resolveStatus,
```
Comment (sorawit) [inline chain/emitter/oracle.go]: I don't need to need special logic branch for result.

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Review by taobun — CHANGES_REQUESTED** [review]
Add change log

**Review by taobun — APPROVED** [review]

**Pull request merged**